File: client/src/api/userService.ts
// =============================================================================
// MTS TELECOM - User Service
// =============================================================================

import api from "./client";
import {
  AdminHardDeleteRequestPayload,
  CreateInternalUserRequest,
  NotificationPreferences,
  PageRequest,
  PageResponse,
  UserResponse,
  UserRole,
} from "../types";

const USERS_PREFIX = "/users";

export const userService = {
  getAllUsers: async (): Promise<UserResponse[]> => {
    const response = await api.get<PageResponse<UserResponse>>(USERS_PREFIX, {
      params: { size: 1000 },
    });
    return response.data.content;
  },

  getUsers: async (page?: PageRequest): Promise<PageResponse<UserResponse>> => {
    const params = new URLSearchParams();
    if (page) {
      if (page.page !== undefined) params.append("page", String(page.page));
      if (page.size !== undefined) params.append("size", String(page.size));
    }
    const response = await api.get<PageResponse<UserResponse>>(USERS_PREFIX, { params });
    return response.data;
  },

  getUserById: async (id: number): Promise<UserResponse> => {
    const response = await api.get<UserResponse>(`${USERS_PREFIX}/${id}`);
    return response.data;
  },

  getUsersByRole: async (role: UserRole): Promise<UserResponse[]> => {
    const response = await api.get<UserResponse[]>(`${USERS_PREFIX}/role/${role}`);
    return response.data;
  },

  getAgents: async (): Promise<UserResponse[]> => {
    const response = await api.get<UserResponse[]>(`${USERS_PREFIX}/agents/available`);
    return response.data;
  },

  createInternalUser: async (data: CreateInternalUserRequest): Promise<UserResponse> => {
    const response = await api.post<UserResponse>(`${USERS_PREFIX}/internal`, data);
    return response.data;
  },

  updateUser: async (
    id: number,
    data: { firstName?: string; lastName?: string; phone?: string },
  ): Promise<UserResponse> => {
    const response = await api.put<UserResponse>(`${USERS_PREFIX}/${id}`, data);
    return response.data;
  },

  changeRole: async (id: number, role: UserRole): Promise<UserResponse> => {
    const response = await api.put<UserResponse>(`${USERS_PREFIX}/${id}/role`, { role });
    return response.data;
  },

  deactivateUser: async (id: number): Promise<void> => {
    await api.post(`${USERS_PREFIX}/${id}/deactivate`);
  },

  activateUser: async (id: number): Promise<void> => {
    await api.post(`${USERS_PREFIX}/${id}/activate`);
  },

  hardDeleteUser: async (id: number, payload: AdminHardDeleteRequestPayload): Promise<void> => {
    await api.delete(`${USERS_PREFIX}/${id}/hard-delete`, { data: payload });
  },

  requestHardDeleteChallenge: async (id: number): Promise<{ message: string }> => {
    const response = await api.post<{ message: string }>(
      `${USERS_PREFIX}/${id}/hard-delete/challenge`,
    );
    return response.data;
  },

  setPasswordByAdmin: async (id: number, newPassword: string): Promise<void> => {
    await api.put(`${USERS_PREFIX}/${id}/password`, { newPassword });
  },

  resetPassword: async (id: number): Promise<void> => {
    await api.post(`${USERS_PREFIX}/${id}/reset-password`, {});
  },

  getProfile: async (): Promise<UserResponse> => {
    const response = await api.get<UserResponse>("/users/me");
    return response.data;
  },

  uploadAvatar: async (file: File): Promise<UserResponse> => {
    const formData = new FormData();
    formData.append("file", file);

    const response = await api.post<UserResponse>("/users/me/avatar", formData, {
      headers: { "Content-Type": "multipart/form-data" },
    });

    return response.data;
  },

  updateProfile: async (data: { fullName?: string; phone?: string }): Promise<UserResponse> => {
    const normalizedPayload: {
      firstName?: string;
      lastName?: string;
      phone?: string;
    } = {
      phone: data.phone,
    };

    if (data.fullName && data.fullName.trim()) {
      const parts = data.fullName.trim().split(/\s+/);
      normalizedPayload.firstName = parts[0];
      normalizedPayload.lastName = parts.slice(1).join(" ") || "-";
    }

    const response = await api.put<UserResponse>("/users/me", normalizedPayload);
    return response.data;
  },

  changePassword: async (data: { currentPassword: string; newPassword: string }): Promise<void> => {
    await api.put("/users/me/change-password", data);
  },

  getNotificationPreferences: async (): Promise<NotificationPreferences> => {
    const response = await api.get<NotificationPreferences>("/users/me/notification-preferences");
    return response.data;
  },

  updateNotificationPreferences: async (prefs: NotificationPreferences): Promise<void> => {
    await api.post("/users/me/notification-preferences", prefs);
  },
};

export default userService;

File: server/src/main/java/com/billcom/mts/service/UserService.java
package com.billcom.mts.service;

import com.billcom.mts.dto.user.UserResponse;
import com.billcom.mts.dto.user.UserUpdateRequest;
import com.billcom.mts.entity.User;
import com.billcom.mts.enums.UserRole;
import org.springframework.core.io.Resource;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

/**
 * Service interface for user management.
 */
public interface UserService {

    /**
     * Get user by ID.
     */
    UserResponse getUserById(Long id);

    /**
     * Get user entity by ID.
     */
    User getUserEntityById(Long id);

    /**
     * Get user by email.
     */
    UserResponse getUserByEmail(String email);

    /**
     * Update user profile.
     */
    UserResponse updateUser(Long id, UserUpdateRequest request);

    /**
     * Update user role (admin only).
     */
    UserResponse updateUserRole(Long userId, UserRole role);

    /**
     * Change user password after validating the current password.
     */
    void changePassword(Long userId, String currentPassword, String newPassword);

    /**
     * Set user password directly as admin.
     */
    void setPasswordByAdmin(Long userId, String newPassword);

    /**
     * Deactivate user account.
     */
    void deactivateUser(Long userId);

    /**
     * Activate user account.
     */
    void activateUser(Long userId);

    /**
     * Permanently delete an internal user account when business dependencies
     * make the operation safe.
     */
    void hardDeleteUserByAdmin(Long userId);

    /**
     * Permanently delete a client account when the related client profile is safe
     * to remove.
     */
    void hardDeleteClientAccountByAdmin(Long userId);

    /**
     * Get all users with pagination.
     */
    Page<UserResponse> getAllUsers(Pageable pageable);

    /**
     * Get users by role.
     */
    List<UserResponse> getUsersByRole(UserRole role);

    /**
     * Get all available agents for ticket assignment.
     */
    List<UserResponse> getAvailableAgents();

    /**
     * Count total users (for Admin dashboard KPI).
     */
    long count();

    /**
     * Update notification preferences (JSON).
     */
    void updateNotificationPreferences(Long userId, java.util.Map<String, Object> preferences);

    /**
     * Get notification preferences (JSON string).
     */
    String getNotificationPreferences(Long userId);

    /**
     * Upload avatar image for user profile.
     */
    UserResponse uploadAvatar(Long userId, MultipartFile file);

    /**
     * Get avatar as downloadable resource.
     */
    Resource getAvatarResource(String filename);

    /**
     * Detect avatar content type.
     */
    String getAvatarContentType(String filename);
}
